File: typescript/tests/giftcard/application/services/bookingResultRefunding.test.ts
import { beforeAll, describe, expect, it } from 'vitest';
import { generateId } from '@/common/domain/identity/entityId.js';
import { Money } from '@/common/domain/primitive/money.js';
import {
  bookingConfirmed,
  bookingRejected,
} from '@/booking/domain/events/bookingResultEvents.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import { BookingResultRefunding } from '@/giftcard/application/services/bookingResultRefunding.js';
import { GiftCardRefunding } from '@/giftcard/application/usecases/giftCardRefunding.js';
import { RefundGiftCardPolicy } from '@/giftcard/application/policies/refundGiftCardPolicy.js';
import { GiftCardId } from '@/giftcard/domain/giftcard/giftCardId.js';
import { SqliteGiftCardRepository } from '@/giftcard/infrastructure/sqliteGiftCardRepository.js';
import { DatabaseSetup } from '../../../testsupport/databaseSetup.js';
import { GiftCardAggregateFactory } from '../../../testsupport/giftcard/aggregateFactory.js';

describe('BookingResultRefunding', () => {
  let repository: SqliteGiftCardRepository;
  let service: BookingResultRefunding;

  beforeAll(() => {
    const database = DatabaseSetup.initializeInMemoryDb('giftcard');
    repository = new SqliteGiftCardRepository(database);
    const policy = new RefundGiftCardPolicy();
    const useCase = new GiftCardRefunding(repository);
    service = new BookingResultRefunding(policy, useCase);
  });

  function createBooking(): BookingId {
    return generateId((value) => new BookingId(value));
  }

  describe('construction', () => {
    it('rejects null parameters', () => {
      const policy = new RefundGiftCardPolicy();
      const useCase = new GiftCardRefunding(repository);

      expect(() => new BookingResultRefunding(null as unknown as RefundGiftCardPolicy, useCase)).toThrow();
      expect(() => new BookingResultRefunding(policy, null as unknown as GiftCardRefunding)).toThrow();
    });
  });

  describe('booking results handling', () => {
    it('on rejected should restore balance', () => {
      const card = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const bookingId = createBooking();
      const refundAmount = new Money(20);

      service.handleBookingResults(bookingRejected(bookingId, card.id().value.value, refundAmount));

      const after = repository.findById(card.id());
      expect(after).not.toBeNull();
      expect(after!.balance()).toEqual(refundAmount);
    });

    it('on confirmed should do nothing', () => {
      const card = GiftCardAggregateFactory.getSavedGiftCard(repository);
      const bookingId = createBooking();
      const amount = new Money(10);

      service.handleBookingResults(bookingConfirmed(bookingId, card.id().value.value, amount));

      const persisted = repository.findById(card.id());
      expect(persisted).not.toBeNull();
      expect(persisted!.balance()).toEqual(Money.zero());
    });

    it('should fail if card does not exist', () => {
      const nonExisting = generateId((value) => new GiftCardId(value));
      const bookingId = createBooking();
      const amount = new Money(10);

      expect(() =>
        service.handleBookingResults(bookingRejected(bookingId, nonExisting.value.value, amount))
      ).toThrow();
    });
  });
});

File: java/src/test/java/it/giannibombelli/wsc2026/giftcard/application/services/BookingResultRefundingTest.java
package it.giannibombelli.wsc2026.giftcard.application.services;

import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.domain.identity.EntityId;
import it.giannibombelli.wsc2026.common.domain.primitive.Money;
import it.giannibombelli.wsc2026.giftcard.application.usecases.GiftCardRefunding;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.application.policies.RefundGiftCardPolicy;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;
import it.giannibombelli.wsc2026.giftcard.infrastructure.SqliteGiftCardRepository;
import it.giannibombelli.wsc2026.testsupport.DatabaseSetup;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import javax.sql.DataSource;
import java.math.BigDecimal;
import java.util.Optional;

import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.createBooking;
import static it.giannibombelli.wsc2026.testsupport.AggregateFactory.getSavedGiftCard;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class BookingResultRefundingTest {

    private GiftCardRepository repository;
    private BookingResultRefunding refundGiftCardService;

    @BeforeEach
    void setUp() {
        final DataSource dataSource = DatabaseSetup.initializeInMemoryDb("giftcard");
        repository = new SqliteGiftCardRepository(dataSource);

        RefundGiftCardPolicy policy = new RefundGiftCardPolicy();
        GiftCardRefunding useCase = new GiftCardRefunding(repository);
        refundGiftCardService = new BookingResultRefunding(policy, useCase);
    }

    @Nested
    class Construction {
        @Test
        void rejectNullParameters() {
            RefundGiftCardPolicy policy = new RefundGiftCardPolicy();
            GiftCardRefunding useCase = new GiftCardRefunding(repository);

            assertThatThrownBy(() -> new BookingResultRefunding(null, useCase))
                .isInstanceOf(NullPointerException.class);

            assertThatThrownBy(() -> new BookingResultRefunding(policy, null))
                .isInstanceOf(NullPointerException.class);
        }
    }

    @Nested
    class BookingResultsHandling {
        @Test
        void onRejected_shouldRestoreBalance() {
            GiftCard card = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money refundAmount = new Money(new BigDecimal("20.00"));

            refundGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingRejected(booking.id(), card.id().value(), refundAmount)
            );

            Optional<GiftCard> after = repository.findById(card.id());
            assertThat(after).isPresent();
            assertThat(after.get().balance()).isEqualTo(refundAmount);
        }

        @Test
        void onConfirmed_shouldDoNothing() {
            GiftCard card = getSavedGiftCard(repository);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("10.00"));

            refundGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingConfirmed(booking.id(), card.id().value(), amount)
            );

            GiftCard persisted = repository.findById(card.id()).orElseThrow();
            assertThat(persisted.balance()).isEqualTo(Money.zero());
        }

        @Test
        void shouldFailIfCardDoesNotExist() {
            GiftCardId nonExisting = EntityId.generate(GiftCardId::new);
            Booking booking = createBooking();
            Money amount = new Money(new BigDecimal("10.00"));

            assertThatThrownBy(() -> refundGiftCardService.handleBookingResults(
                new BookingResultEvents.BookingRejected(booking.id(), nonExisting.value(), amount)
            ))
                .isInstanceOf(IllegalStateException.class);
        }
    }
}
